files/docker/node/addons/healthcheck.sh

```diff
@@ -1,43 +1,132 @@
-#!/usr/bin/env bash
+#!/bin/bash
+# shellcheck source=/dev/null
+#
+######################################
+# User Variables - Change as desired #
+# Common variables set in env file   #
+######################################
 
-source /opt/cardano/cnode/scripts/env
+ENTRYPOINT_PROCESS="${ENTRYPOINT_PROCESS:-cnode.sh}"        # Get the script from ENTRYPOINT_PROCESS or default to "cnode.sh" if not set
+CPU_THRESHOLD="${CPU_THRESHOLD:-80}"                        # The CPU threshold to warn about if the sidecar process exceeds this for more than 60 seconds, defaults to 80%
+RETRIES="${RETRIES:-20}"                                    # The number of retries if tip is not incrementing, or cpu usage is over the threshold
 
-CCLI=$(which cardano-cli)
+######################################
+# Do NOT modify code below           #
+######################################
 
-if [[ "$NETWORK" == "guild-mainnet" ]]; then NETWORK=mainnet; fi
+if [[ "${ENTRYPOINT_PROCESS}" == "cnode.sh" ]]; then
+    source /opt/cardano/cnode/scripts/env
+else
+    # Source in offline mode for sidecar helper scripts
+    source /opt/cardano/cnode/scripts/env offline
+fi
 
-# For querying tip, the seperation of testnet-magic vs mainnet as argument is optional
+# Define a mapping of scripts to their corresponding binaries, when defined check the binary is running and its CPU usage instead of the wrapper script.
+declare -A SCRIPT_TO_BINARY_MAP
+SCRIPT_TO_BINARY_MAP=(
+    ["cncli.sh"]="cncli"
+    ["mithril-signer.sh"]="mithril-signer"
+)
 
-FIRST=$($CCLI query tip --testnet-magic ${NWMAGIC} | jq .block)
+# Define scripts which may sleep between executions of the binary.
+SLEEPING_SCRIPTS=("cncli.sh")
 
-if [[ "${ENABLE_KOIOS}" == "N" ]] || [[ -z "${KOIOS_API}" ]]; then
-    # when KOIOS is not enabled or KOIOS_API is unset, use default behavior
-    sleep 60;
-    SECOND=$($CCLI query tip --testnet-magic ${NWMAGIC} | jq .block)
-    if [[ "$FIRST" -ge "$SECOND" ]]; then
-        echo "there is a problem"
-        exit 1
+# Function to check if a process is running and its CPU usage
+check_process() {
+    local process_name="$1"
+    local cpu_threshold="$2"
+
+    for (( CHECK=1; CHECK<=RETRIES; CHECK++ )); do
```

Comment:
```suggestion
    for (( CHECK=0; CHECK<=RETRIES; CHECK++ )); do
```
Suggestion: Modify the loop condition to handle RETRIES=0. 
- With this...
  -  'RETRIES=3' results in the loop running a total of 4 times
  -  'RETRIES=0' results in the loop running a total of 1 times